### PR TITLE
Docking syncs

### PR DESCRIPTION
**Pull request by ahk1221**
Docking/undocking states synced/saved on the server, undocking now updates correctly on all clients, though there still is an issue with the animation. Maybe related to movement packets being suppressed?

### CODE REVIEW

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 1]
```diff
@@ -438,6 +428,62 @@ public void CreateRemotePlayers(object sender, EventArgs eventArgs)
...
+            public void CreateVehicles(object sender, EventArgs eventArgs)
+            {
+                ThrottledBuilder.main.QueueDrained -= CreateVehicles;
+
+                List<VehicleModel> cyclopses = vehicleModels
+                    .Where(x => x.TechType == TechType.Cyclops)
+                    .ToList();
+
+                List<VehicleModel> nonCyclopses = vehicleModels
+                    .Where(x => !cyclopses.Any(x2 => x2.TechType == x2.TechType))
```
Comment (Measurity) [inline NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: I think there's a bug here: `x2.TechType == x2.TechType` should be `x.TechType == x2.TechType`

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 2]
```diff
@@ -438,6 +428,62 @@ public void CreateRemotePlayers(object sender, EventArgs eventArgs)
...
+                    .ToList();
+
+                foreach(VehicleModel vehicle in cyclopses)
+                {
+                    vehicles.CreateVehicle(vehicle);
```
Comment (Measurity) [inline NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: If both nonCyclopses and Cyclopses need to call `CreateVehicle` then why not just iterate on `vehicleModels` without filtering for `nonCyclopses` and `cyclopses`?

Comment (ahk1221) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: This is to spawn the cyclopses first, as a non-cyclops could be docked inside a cyclops.

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 3]
```diff
@@ -237,6 +238,15 @@ private void SpawnVehicles(List<VehicleModel> vehicleModels)
...
 
+            List<VehicleModel> cyclopses = vehicleModels
+                .Where(x => x.TechType == TechType.Cyclops)
+                .ToList();
+
+            foreach(VehicleModel cyclops in cyclopses)
+            {
+                vehicles.CreateVehicle(cyclops);
+            }
+
```
[the remark this answers is not in the document]
Comment (ahk1221) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: This is to spawn the cyclopses first, as a non-cyclops could be docked inside a cyclops.

File: NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            Optional<VehicleModel> vehicle = vehicleData.GetVehicleModel(packet.VehicleGuid);
+            if (!vehicle.IsPresent())
+            {
+                return;
+            }
+
+            VehicleModel vehicleModel = vehicle.Get();
+            vehicleModel.DockingBayGuid = Optional<string>.Of(packet.DockGuid);
+
+            //playerManager.SendPacketToOtherPlayers(packet, player);
```
[the remark this answers is not in the document]
Comment (ahk1221) [reply NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs]: They don't need to, as the client would automatically dock the vehicle once it enters the collider.

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs (hunk 1 above)
Comment (ahk1221) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Oops, yeah you are right. Will fix.

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs (hunk 2 above)
Comment (alexanderheuts) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Leave some code comments to explain this for future reference.

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexanderheuts) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Leave some code comments.

File: NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (alexanderheuts) [reply NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs]: Remove the line and add a comment to explain this

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 5]
```diff
@@ -438,6 +448,62 @@ public void CreateRemotePlayers(object sender, EventArgs eventArgs)
             }
         }
 
+        private void SpawnVehiclesAfterBasePiecesFinish(List<VehicleModel> vehicleModels, bool basePiecesToSpawn)
+        {
+            Log.Info("Received initial sync packet with " + vehicleModels.Count + " vehicles.");
```
[the remark this answers is not in the document]
Comment (ahk1221) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Will do.

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs (hunk 2 above)
Comment (Sunrunner37) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Do we need to leave frames between spawning cyclops and non-cyclops? From my understanding, cyclops is usually spawned by including a unity scene that performs the build animation (even a brief one when `sub` is used)

File: NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            this.playerManager = playerManager;
+            this.vehicleData = vehicleData;
+        }
+
+        public override void Process(VehicleDocking packet, Player player)
+        {
+            Optional<VehicleModel> vehicle = vehicleData.GetVehicleModel(packet.VehicleGuid);
+            if (!vehicle.IsPresent())
+            {
```
Comment (Sunrunner37) [inline NitroxServer/Communication/Packets/Processors/VehicleDockingProcessor.cs]: Should we log an error?

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 7]
```diff
@@ -438,6 +428,62 @@ public void CreateRemotePlayers(object sender, EventArgs eventArgs)
...
+
+            public void CreateVehicles(object sender, EventArgs eventArgs)
+            {
+                ThrottledBuilder.main.QueueDrained -= CreateVehicles;
+
+                List<VehicleModel> cyclopses = vehicleModels
+                    .Where(x => x.TechType == TechType.Cyclops)
+                    .ToList();
+
+                List<VehicleModel> nonCyclopses = vehicleModels
```
Comment (Sunrunner37) [inline NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Might also be cleaner to create two small methods at the bottom of this class to encapsulate the filtering and the looping:

```
// Detailed comment on why these are two methods 
spawnCyclopsVehicles(vehicleModels);
spawnOtherVehicles(vehicModels);
```

File: NitroxServer/Communication/Packets/Processors/VehicleUndockingProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            this.playerManager = playerManager;
+            this.vehicleData = vehicleData;
+        }
+
+        public override void Process(VehicleUndocking packet, Player player)
+        {
+            Optional<VehicleModel> vehicle = vehicleData.GetVehicleModel(packet.VehicleGuid);
+            if (!vehicle.IsPresent())
+            {
+                return;
```
Comment (Sunrunner37) [inline NitroxServer/Communication/Packets/Processors/VehicleUndockingProcessor.cs]: Log error here as well?

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs (hunk 2 above)
Comment (MarijnS95) [reply NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: Is it necessary to create two in-memory lists (`ToList()`) of these two queries, or can they be left as `IEnumerable`s that do the filtering when they are iterated?

File: NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs [hunk 9]
```diff
@@ -64,6 +64,7 @@ public override void Process(InitialPlayerSync packet)
 
             SpawnInventoryItemsAfterBasePiecesFinish(packet.InventoryItems, hasBasePiecesToSpawn, packet.PlayerGuid);
             SpawnRemotePlayersAfterBasePiecesFinish(packet.RemotePlayerData, hasBasePiecesToSpawn);
+            SpawnVehiclesAfterBasePiecesFinish(packet.Vehicles, hasBasePiecesToSpawn);
```
Comment (alexanderheuts) [inline NitroxClient/Communication/Packets/Processors/InitialPlayerSyncProcessor.cs]: There's going to be an inherent problem. Cyclops furniture is a basePiece. Spawning cyclops furniture needs to happen after spawning a cyclops.

**Pull request merged**